jupyter_server/services/contents/fileidmanager.py

```diff
@@ -0,0 +1,411 @@
+import os
+import sqlite3
+import stat
+from typing import Optional
+
+from jupyter_core.paths import jupyter_data_dir
+from traitlets import TraitError, Unicode, validate
+from traitlets.config.configurable import LoggingConfigurable
+
+
+class StatStruct:
+    ino: int
+    crtime: Optional[int]
+    mtime: int
+    is_dir: bool
+
+
+class FileIdManager(LoggingConfigurable):
+    """
+    Manager that supports tracks files across their lifetime by associating
+    each with a unique file ID, which is maintained across filesystem operations.
+
+    Notes
+    -----
+
+    All private helper methods prefixed with an underscore (except `__init__()`)
+    do NOT commit their SQL statements in a transaction via `self.con.commit()`.
+    This responsibility is delegated to the public method calling them to
+    increase performance. Committing multiple SQL transactions in serial is much
+    slower than committing a single SQL transaction wrapping all SQL statements
+    performed during a method's procedure body.
+    """
+
+    root_dir = Unicode(
+        help=("The root being served by Jupyter server. Must be an absolute path."), config=True
+    )
+
+    db_path = Unicode(
+        default_value=os.path.join(jupyter_data_dir(), "file_id_manager.db"),
+        help=(
+            "The path of the DB file used by `FileIdManager`. "
+            "Defaults to `jupyter_data_dir()/file_id_manager.db`."
+        ),
+        config=True,
+    )
+
+    def __init__(self, **kwargs):
+        # pass args and kwargs to parent Configurable
+        super().__init__(**kwargs)
+        # initialize connection with db
+        self.con = sqlite3.connect(self.db_path)
+        self.log.debug("FileIdManager : Creating File ID tables and indices")
+        self.con.execute(
+            "CREATE TABLE IF NOT EXISTS Files("
+            "id INTEGER PRIMARY KEY AUTOINCREMENT, "
+            # uniqueness constraint relaxed here because we need to keep records
+            # of deleted files which may occupy same path
+            "path TEXT NOT NULL, "
+            "ino INTEGER NOT NULL UNIQUE, "
```

Comment:
Inode (`st_ino`) [is not unique by itself](https://docs.python.org/3/library/os.html#os.stat_result.st_ino) and needs to be coupled with `st_dev` to gain uniqueness within a set of devices managed by a given server.  I also think we will have issues when mounted file systems are in play (e.g., I think I recall that the inode reflects that of the NFS server for example, not necessarily the client mounting the filesystem, besides other idiosyncrasies that will creep in as these kinds of things vary with OS), not to mention non-file systems like Databases or Cloud Object Storage, etc.  So this feels like a bit of a slippery slope, IMHO.

Reply:
Ah... right. Good catch. I'll see what I can do about this. I think by default, the File ID Manager will only support local filesystems and NFS. For other cloud services, it's possible that they already expose a file ID, so the custom File ID manager implementation should be straightforward.

Reply:
> For other cloud services, it's possible that they already expose a file ID, so the custom File ID manager implementation should be straightforward.

I believe this is saying that non-file system-based `ContentsManager` derivations must implement their own `FileIdManager`.  Is that correct?

Regarding NFS, I wouldn't be surprised if the `st_dev` value changes if a previously mounted FS is unmounted, then remounted (following the mount of others or even a reboot).  (This [hunch may be confirmed](https://utcc.utoronto.ca/~cks/space/blog/unix/NFSMultiFSFault) if I'm reading things correctly.)

Reply:
> I believe this is saying that non-file system-based ContentsManager derivations must implement their own FileIdManager. Is that correct?

If this is the case (prolly it is), the FileIdManagement implementation should be configurable (like the ContentManager...). It may already be the case (have not looked at the code, sorry).

Reply:
@echarles - this is discussed here: https://github.com/jupyter-server/jupyter_server/pull/921#discussion_r936123139 although I don't think we should require that.